server/webapp/WEB-INF/rails/webpack/models/pipeline_configs/job.ts

```diff
@@ -15,31 +15,41 @@
  */
 
 import JsonUtils from "helpers/json_utils";
-import {Stream} from "mithril/stream";
+import * as _ from "lodash";
 import * as stream from "mithril/stream";
+import {Stream} from "mithril/stream";
 import {ValidatableMixin} from "models/mixins/new_validatable_mixin";
+import {EnvironmentVariableConfig} from "models/pipeline_configs/environment_variable_config";
 import {Task} from "models/pipeline_configs/task";
+import * as s from "underscore.string";
 
 export class Job extends ValidatableMixin {
   name: Stream<string>;
+  environmentVariables: Stream<EnvironmentVariableConfig[]>;
   tasks: Stream<Task[]>;
 
-  constructor(name: string, tasks: Task[]) {
+  constructor(name: string, tasks: Task[], envVars: EnvironmentVariableConfig[]) {
     super();
 
     ValidatableMixin.call(this);
     this.name = stream(name);
     this.tasks = stream(tasks);
+    this.environmentVariables = stream(envVars);
     this.validatePresenceOf("name");
     this.validateIdFormat("name");
 
     this.validatePresenceOf("tasks");
     this.validateNonEmptyCollection("tasks", {message: "A job must have at least one task"});
     this.validateEach("tasks");
+    this.validateEach("environmentVariables");
   }
 
   toApiPayload() {
-    return JsonUtils.toSnakeCasedObject(this);
+    const raw = JsonUtils.toSnakeCasedObject(this);
+    raw.environment_variables = _.filter(this.environmentVariables(), (variable) => {
```

Comment:
Instead, shouldn't we rely on the validation that an `EnvironmentVariableConfig` cannot have a blank name? We _shouldn't_ have to do this, unless I'm missing something nuanced.

Reply:
Yeah, I'm not a fan of doing this either. The reason it's necessary is because when have 2 env vars that are on the page when it loads (one empty secure and one empty plain). They don't have to fill them in to submit the form.

Reply:
Those vars would be considered invalid if we validate the presence of the name. I couldn't think of a good way to have those who vars be on the page on load, and have the validations be a bit more standard. 

Does that makes sense?

Reply:
@ibnc I know we chatted on slack, but going to copy responses here for documentation.

I understand the reasoning now.

What I would opt for is to create a `stream` to hold unsaved values and have the component convert them into `EnvironmentVariableConfig` objects `onchange`.

That way you can have the right validations on environment variables, and not have to worry about this case where we leave the form unfilled while validating `EnvironmentVariableConfig` instances.

Kind of analogous to how the task editor starts with an empty task array and only creates Task instances after typing some text and pressing `enter`.